pkg/standalone-utils/test/GaugeActions.test.ts

```diff
@@ -0,0 +1,659 @@
+        itDepositsWithRefsAndAmounts(totalLpTokens);
+      });
+
+      context('when depositing 0 tokens', () => {
+        itDepositsWithRefsAndAmounts(bn(0));
+      });
+    }
+
+    function itDepositsWithRefsAndAmounts(amount: BigNumber) {
+      context('when using immediate amounts', () => {
+        itDepositsTokens(amount, amount);
+      });
+
+      context('when using chained references', () => {
+        const reference = toChainedReference(123);
+
+        sharedBeforeEach('set chained reference', async () => {
+          await setChainedReferenceContents(relayer, reference, amount);
+        });
+
+        itDepositsTokens(reference, amount);
+      });
+    }
+
+    function itDepositsTokens(depositAmountOrRef: BigNumber, expectedAmount: BigNumber) {
+      let receipt: ContractReceipt;
+      let tokenSenderAddress: string;
+      let tokenRecipientAddress: string;
+
+      sharedBeforeEach('get addresses', async () => {
+        tokenSenderAddress = TypesConverter.toAddress(tokenSender);
+        tokenRecipientAddress = TypesConverter.toAddress(tokenRecipient);
+      });
+
+      sharedBeforeEach('check initial balances and make deposit', async () => {
+        // Start: sender has all the BPT, no gauge tokens minted, recipient is clean unless it is sender.
+        expect(await lpToken.balanceOf(gauge.address)).to.be.eq(0);
+        expect(await lpToken.balanceOf(tokenSenderAddress)).to.be.eq(totalLpTokens);
+        if (tokenSenderAddress != tokenRecipientAddress) {
+          expect(await lpToken.balanceOf(tokenRecipientAddress)).to.be.eq(0);
+        }
+
+        expect(await gauge.balanceOf(gauge.address)).to.be.eq(0);
+        expect(await gauge.balanceOf(tokenSenderAddress)).to.be.eq(0);
+        expect(await gauge.balanceOf(tokenRecipientAddress)).to.be.eq(0);
+
+        const tx = await relayer.connect(userSender).multicall([
+          encodeDeposit({
+            gauge: gauge,
+            sender: tokenSenderAddress,
+            recipient: tokenRecipientAddress,
+            amount: depositAmountOrRef,
+          }),
+        ]);
+        receipt = await tx.wait();
+      });
+
+      // Short-circuit when no tokens are deposited.
+      expectedAmount.gt(0) &&
+        it('pulls BPT tokens from sender if necessary', async () => {
+          // This transfer is skipped if token sender is relayer, but addresses are undefined outside this scope.
+          if (tokenSenderAddress != relayer.address) {
+            expectTransferEvent(
+              receipt,
+              { from: tokenSenderAddress, to: relayer.address, value: expectedAmount },
+              lpToken.address
+            );
+          }
+        });
+
+      it("approves gauge to use relayer's BPT funds", async () => {
+        expectEvent.inIndirectReceipt(receipt, lpToken.interface, 'Approval', {
+          owner: relayer.address,
+          spender: gauge.address,
+          value: expectedAmount,
+        });
+      });
```

Comment:
This test is not _wrong_, but the fact that this happens is a bit of an implementation detail on the relayer. This approval doesn't _need_ to happen (maybe the relayer sets e.g. infite approval at some other point or whatever). We can keep this now, but note that what is important is that the deposit succeeds - how we get there doesn't matter as much